galsim/utilities.py

```diff
@@ -296,6 +296,25 @@ def _convertPositions(pos, units, func):
 
     return pos
 
+def _lin_approx_err(x, f, i):
+    """Error as \int abs(f(x) - approx(x)) when using ith data point to make piecewise linear
+    approximation."""
+    xleft, xright = x[:i+1], x[i:]
+    fleft, fright = f[:i+1], f[i:]
+    xi, fi = x[i], f[i]
+    mleft = (fi-f[0])/(xi-x[0])
+    mright = (f[-1]-fi)/(x[-1]-xi)
+    f2left = f[0]+mleft*(xleft-x[0])
+    f2right = fi+mright*(xright-xi)
+    return np.trapz(np.abs(fleft-f2left), xleft), np.trapz(np.abs(fright-f2right), xright)
+
+def _lin_approx_split(x, f):
+    """Optimally split a tabulated function into two-part piecewise linear approximation.
+    """
+    errs = [_lin_approx_err(x, f, i) for i in xrange(1, len(x)-1)]
```

Comment:
It's a little messy, but the L2 split point is definitely doable in O(N).  Here's some math (I haven't tried to code any of this up yet).  We have a bunch of point (xj,fj), and we want to find the point i where the error is minimized.  This is the sum of errors for 0<j<i and i<j<N  (where N is the last point, aka -1).

The linear approximations are:

g_left(x) = f0 + (fi-f0)/(xi-x0) * (x-x0)
g_right(x) = fN + (fi-fN)/(xi-xN) * (x-xN)

The left-side L2 error is:

err_left = sum_0<j<i (fj - gleft(xj))^2
    = sum_0<j<i (fj - f0)^2
       + 2(fi-f0)/(xi-x0) sum_0<j<i (fj-f0)(xj-x0)
       + (fi-f0)^2/(xi-x0)^2 sum_0<j<i (xj-x0)^2

The right-side error is the same, but sum_i<j<N and x0->xN, f0->fN.

These sums can all be efficiently computed with `numpy.cumsum`, so I think everything is O(N) and vectorizable, so it should be pretty fast.  Do you want to try coding it up?  Or would you like me to give it a go?